contracts/RadiCards.sol

```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
-    uint256 tokenId = tokenIdPointer;
-
-    super._mint(to, tokenId);
-    _setTokenURI(tokenId, tokenURI);
-
-    tokenIdPointer = tokenIdPointer.add(1);
-
-    return tokenId;
-  }
-
-  function burn(uint256 _tokenId) public  {
-    revert("Radi.Cards are censorship resistant!");
-  }
-
-  function tokenURI(uint256 _tokenId) public view returns (string) {
-    require(exists(_tokenId));
-
-    return Strings.strConcat(tokenBaseURI, tokenURIs[_tokenId]);
-  }
-
-  function tokenDetails(uint256 _tokenId)
-  public view
-  returns (
-    address _gifter,
-    uint256 _giftingAmount,
-    string _message,
-    string _extra,
-    uint256 _cardIndex,
-    uint256 _benefactorIndex
-  ) {
-    require(exists(_tokenId));
-    RadiCard memory _radiCard = tokenIdToRadiCardIndex[_tokenId];
-    return (
-      _radiCard.gifter,
-      _radiCard.giftingAmount,
-      _radiCard.message,
-      _radiCard.extra,
-      _radiCard.cardIndex,
-      _radiCard.benefactorIndex
+contract RadiCards is ERC721Token, ERC721Holder, Whitelist {
+    using SafeMath for uint256;
+
+    // dai support
+    StandardToken daiContract;
+    Medianizer medianizerContract;
+
+    string public tokenBaseURI = "https://ipfs.infura.io/ipfs/";
+    uint256 public tokenIdPointer = 0;
+
+    struct Benefactor {
+        address ethAddress;
+        string name;
+        string website;
+        string logo;
+    }
+
+    struct CardDesign {
+        string tokenURI;
+        bool active;
+        uint minted;
+        uint maxQnty; //set to zero for unlimited
+        //minimum price per card set in Atto (SI prefix for 1x10-18 dai)
+        uint256 minPrice; //set to zero to default to the minimumContribution
+    }
+
+    // claimable link support
+    enum Statuses { Empty, Deposited, Claimed, Cancelled }
+    uint256 public EPHEMERAL_ADDRESS_FEE = 0.01 ether;
+    mapping(address => uint256) public ephemeralWalletCards; // ephemeral wallet => tokenId
+
+    struct RadiCard {
+        address gifter;
+        string message;
+        bool daiDonation;
+        uint256 giftAmount;
+        uint256 donationAmount;
+        Statuses status;
+        uint256 cardIndex;
+        uint256 benefactorIndex;
+    }
+
+    mapping(uint256 => Benefactor) public benefactors;
+    uint256[] internal benefactorsIndex;
+
+    mapping(uint256 => CardDesign) public cards;
+    uint256[] internal cardsIndex;
+
+    mapping(uint256 => RadiCard) public tokenIdToRadiCardIndex;
+
+    //total gifted/donated in ETH
+    uint256 public totalGiftedInWei;
+    uint256 public totalDonatedInWei;
+
+    //total gifted/donated in DAI
+    uint256 public totalGiftedInAtto; //SI prefix for 1x10-18 dai is Atto.
+    uint256 public totalDonatedInAtto;
+
+    event CardGifted(
+        address indexed _to,
+        uint256 indexed _benefactorIndex,
+        uint256 indexed _cardIndex,
+        address _from,
+        uint256 _tokenId,
+        Statuses status
```

Comment:
Is there a need to understand if the card was gift with `DAI` or `ETH`, if so then maybe add a flag to the event.

Reply:
I agree. I had this originally but the event was too big, with too many params and was failing. will look into optimizing it

Reply:
there is a need as this defines what the units are for the `donationAmount` and `giftAmount`. the same variable will either store Wei or Atto for Eth and Dai respectively. so yes, I agree that we should add this to the event